### PR TITLE
[Zen2] Gather votes from all nodes

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we accept that some nodes may vote for the wrong master in an election.
This is mostly fine because they do end up joining the correct master in the
end, but the lack of a vote from every follower may prevent a future desirable
reconfiguration from taking place.

The solution is to hold another election in a yet-higher term in order to
collect a complete set of votes. Elections are somewhat disruptive so we should
think carefully about when this election should take place. One option is to
wait as late as possible (on the grounds that it might not ever be necessary).
This unfortunately makes it harder to predict how an
apparently-smoothly-running cluster will react to nodes leaving and joining.
Instead we prefer to perform the election as soon as possible in the leader's
term, adding "votes from all followers" to the invariants that we expect to
hold in a stable cluster. The start of a leader's term is already a somewhat
disrupted time for the cluster, so performing another election at this point
does not materially change the cluster's behaviour.

This change implements the logic needed to trigger a new election in order to
satisfy this extra stabilisation condition.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 1]
```diff
@@ -753,6 +766,11 @@ public String toString() {
         private final AckListener ackListener;
         private final ActionListener<Void> publishListener;
 
+        // We may not have accepted our own state before receiving a join from another node, causing its join to be rejected (we cannot
+        // safely accept a join whose last-accepted term/version is ahead of ours), so store them up and process them at the end.
+        // TODO this is unpleasant, is there a better way?
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: Maybe it's not so bad. WDYT?

**Comment by DaveCTurner** [comment]
Iterated test runs found some interesting cases here (I thought I was done at 2c98dd7, all the subsequent commits were chasing test failures) but they've now passed 600 successes in a row on 5d85715bc14087fd3df30371104a5ff3b33ac9b0 so I think this is good to go.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java [hunk 2]
```diff
@@ -222,8 +222,10 @@ PublishWithJoinResponse handlePublishRequest(PublishRequest publishRequest) {
             ensureTermAtLeast(sourceNode, publishRequest.getAcceptedState().term());
             final PublishResponse publishResponse = coordinationState.get().handlePublishRequest(publishRequest);
 
-            if (sourceNode.equals(getLocalNode()) == false) {
-                becomeFollower("handlePublishRequest", sourceNode);
+            if (sourceNode.equals(getLocalNode())) {
+                preVoteCollector.update(getPreVoteResponse(), getLocalNode());
+            } else {
+                becomeFollower("handlePublishRequest", sourceNode); // updates preVoteCollector
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: maybe change comment to "also updates preVoteCollector" to make it clearer that that is not the only purpose (or maybe I just misinterpreted this)

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java (hunk 1 above)
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java]: as reconfiguration (which cares about the joins) can only happen in the next cluster state update (and is only triggered at the end of this publication), I think this is ok.

**Review by ywelsch — APPROVED** [review]
LGTM

**Pull request merged**